### PR TITLE
OMP Speedups

### PR DESCRIPTION
**Pull request by BrettHemes-3M**
Use OpenMP to run joint solution and edge weight calculations across multiple cores.

### CODE REVIEW

File: descartes_planner/src/planning_graph.cpp [hunk 1]
```diff
@@ -814,45 +814,71 @@ bool PlanningGraph::calculateJointSolutions(const std::vector<TrajectoryPtPtr>&
...
-    }
-
-    poses[i].reserve(joint_poses.size());
-    for (auto& sol : joint_poses)
+    #pragma omp for
+    for (std::size_t i = 0; i < points.size(); ++i)
     {
-      poses[i].emplace_back(std::move(sol), points[i]->getTiming());
+      #pragma omp flush (success)
+      if (success) // stop all threads after first observed failure
```
Comment (davetcoleman) [inline descartes_planner/src/planning_graph.cpp:824]: this should be a for loop break if success == false

File: descartes_planner/src/planning_graph.cpp [hunk 2]
```diff
@@ -814,45 +814,71 @@ bool PlanningGraph::calculateJointSolutions(const std::vector<TrajectoryPtPtr>&
...
+  bool success = true;
+  #pragma omp parallel for
   for (std::size_t i = 1; i < poses.size(); ++i)
   {
-    const std::vector<JointTrajectoryPt>& from = poses[i - 1];
-    const std::vector<JointTrajectoryPt>& to = poses[i];
-
-    if (!calculateEdgeWeights(from, to, edges))
+    #pragma omp flush (success)
+    if (success)
```
Comment (davetcoleman) [inline descartes_planner/src/planning_graph.cpp:860]: same here

**Review by davetcoleman — CHANGES_REQUESTED** [review]
Slick addition!

File: descartes_planner/src/planning_graph.cpp (hunk 1 above)
Comment (BrettHemes-3M) [reply descartes_planner/src/planning_graph.cpp:824]: That was my initial thought (and attempt actually) but OpenMP apparently does not allow breaks in parallel for loops.  My thought was that, if I still have to loop at least don't do anything time consuming...  which led to the above implementation.

Comment (BrettHemes-3M) [reply descartes_planner/src/planning_graph.cpp:824]: Also, a failure at this point means you most likely have problems elsewhere.  That said, a few extra for-loop iterations aren't going to matter in the grand scheme of things.

Comment (BrettHemes-3M) [reply descartes_planner/src/planning_graph.cpp:824]: Actually... after reading quick, it seems OpenMP 4.0 added a `cancel` directive (see page 113 [here](http://www.openmp.org/wp-content/uploads/OpenMP_Examples_4.0.1.pdf) for example).  I will check if this works.

**Comment by BrettHemes-3M** [comment]
Closing this as it is superseded by #207

**Pull request closed without merging**